File: src/MatrixCutting.java
// https://code.google.com/codejam/contest/3254486/dashboard#s=p2&a=2
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Scanner;


public class MatrixCutting {

    public static void main(String[] args) {
        Scanner in = new Scanner(new BufferedReader(new InputStreamReader(System.in)));
        int T = in.nextInt();
        for (int i = 1; i <= T; ++i) {
            int N = in.nextInt();
            int M = in.nextInt();
            int[][] matrix = new int[N][M];
            for (int j = 0; j < N; ++j) {
                for (int k = 0; k < M; ++k) {
                    matrix[j][k] = in.nextInt();
                }
            }
            System.out.println("Case #" + i + ": " + solve(matrix));
        }

    }

    private static int solve(int[][] matrix) {
        int N = matrix.length;
        int M = matrix[0].length;
        int[][][][][] cache = new int[M][M + 1][N][N + 1][2];
        return cut(matrix, 0, matrix[0].length, 0, matrix.length, cache)[0];
    }

    private static int[] cut(int[][] matrix, int left, int right, int top, int bottom, int[][][][][] cache) {
        if (cache[left][right][top][bottom][1] > 0) {
            return cache[left][right][top][bottom];
        }

        if (left + 1 == right && top + 1 == bottom) {
            int[] res = {0, matrix[top][left]};
            cache[left][right][top][bottom] = res;
            return res;
        }
        int maxTotal = 0;
        int overallMin = -1;
        for (int i = left + 1; i < right; ++i) {
            int[] leftRes = cut(matrix, left, i, top, bottom, cache);
            int[] rightRes = cut(matrix, i, right, top, bottom, cache);
            overallMin = leftRes[1] < rightRes[1] ? leftRes[1] : rightRes[1];
            int total = overallMin + leftRes[0] + rightRes[0];
            maxTotal = maxTotal > total ? maxTotal : total;
        }
        for (int i = top + 1; i < bottom; ++i) {
            int[] topRes = cut(matrix, left, right, top, i, cache);
            int[] bottomRes = cut(matrix, left, right, i, bottom, cache);
            overallMin = topRes[1] < bottomRes[1] ? topRes[1] : bottomRes[1];
            int total = overallMin + topRes[0] + bottomRes[0];
            maxTotal = maxTotal > total ? maxTotal : total;
        }
        int[] res = {maxTotal, overallMin};
        cache[left][right][top][bottom] = res;
        return res;
    }

}

File: src/matrix_cutting.py
"""https://code.google.com/codejam/contest/3254486/dashboard#s=p2&a=2
Note that, this program will exceed the time limit on large input. But this algorithm is
correct. In order not to time out, you can use the Java implementation of this algorithm
(MatrixCutting.java)."""
from functools import lru_cache


matrix = None


def main():
    global matrix
    T = int(input())
    for i in range(1, T + 1):
        N, M = (int(s) for s in input().split(" "))
        matrix = tuple(tuple(int(s) for s in input().split(" ")) for _ in range(N))
        print("Case #{}: {}".format(i, solve()))


def solve():
    res = cut(0, len(matrix[0]), 0, len(matrix))[0]
    cut.cache_clear()
    return res


@lru_cache(maxsize=None)
def cut(left, right, top, bottom):
    if left + 1 == right and top + 1 == bottom:
        return 0, matrix[top][left]
    max_total = 0
    overall_min = None
    for i in range(left + 1, right):
        left_total, left_min = cut(left, i, top, bottom)
        right_total, right_min = cut(i, right, top, bottom)
        overall_min = min(left_min, right_min)
        total = overall_min + left_total + right_total
        max_total = max(max_total, total)

    for i in range(top + 1, bottom):
        top_total, top_min = cut(left, right, top, i)
        bottom_total, bottom_min = cut(left, right, i, bottom)
        overall_min = min(top_min, bottom_min)
        total = overall_min + top_total + bottom_total
        max_total = max(max_total, total)

    return max_total, overall_min


if __name__ == '__main__':
    main()
